external-crates/move/move-compiler/src/typing/translate.rs

```diff
@@ -2265,3 +2275,57 @@ fn make_arg_types<S: std::fmt::Display, F: Fn() -> S>(
     }
     given
 }
+
+//**************************************************************************************************
+// Module-wide warnings
+//**************************************************************************************************
+
+/// Generates warnings for unused struct types and unused (private) functions.
+fn gen_unused_warnings(context: &mut Context, ident: ModuleIdent, mdef: &T::ModuleDefinition) {
+    if context.compiled_module_names.contains(&ident) {
+        // generate warnings only for modules compiled in this pass rather than for all modules
+        // including pre-compiled libraries for which we do not have the right type of data
+        // collected (as they are not actually compiled here)
+        context
+            .env
+            .add_warning_filter_scope(mdef.warning_filter.clone());
+
+        for (loc, name, fun) in &mdef.functions {
+            context
+                .env
+                .add_warning_filter_scope(fun.warning_filter.clone());
+            if !context.called_fns.contains(name)
+                && fun.entry.is_none()
+                && matches!(fun.visibility, Visibility::Internal)
+            {
+                // TODO: postponing handling of friend functions until we decide what to do with them
+                // vis-a-vis ideas around package-private
+                let msg = format!(
+                    "The non-'public', non-'entry' function '{name}' is never called. \
+                           Consider removing it."
+                );
+                context
+                    .env
+                    .add_diag(diag!(UnusedItem::Function, (loc, msg)))
+            }
+            context.env.pop_warning_filter_scope();
+        }
+
+        for (loc, name, struct_def) in &mdef.structs {
+            context
+                .env
+                .add_warning_filter_scope(struct_def.warning_filter.clone());
```

Comment:
I'm still feeling like we should strip this out, but maybe it is worth bringing to the team for a 3rd/4th/5th opinion

Reply:
Errr actually I definitely think we need to strip this out. Consider this program:
```
module example::m {
    struct TAG1 {}
    struct TAG2 {}
    struct Key<phantom T> has copy, drop store {}

    public fun add_key_1<V: store>(obj: &mut UID, v: V) {
        dynamic_field::add(obj, Key<TAG1> {}, v);
    }
    
    public fun add_key_2<V: store>(obj: &mut UID, v: V) {
        dynamic_field::add(obj, Key<TAG2> {}, v);
    }
}
```
we will complain about `TAG1` and `TAG2` not being packed/used, when they are used for type driven programming.